# **Operators and Types**

In this lesson you'll further explore how Python represents numbers, text, how to convert between them, and some useful operators and methods.


### **Before You Begin**
<div class="alert alert-block alert-info">

**REMINDER:** This file is an executable Jupyter-style notebook. It's made of *cells* — pieces of text or code you can run one at a time.

**Quick run & edit helpers**

* Click a cell to edit it, then press <kbd style="background-color: white; color: black; white-space: nowrap;">⇧ Shift</kbd> + <kbd style="background-color: white; color: black; white-space: nowrap;">⏎ Enter</kbd> to run the cell.
* You can also use the ▶️ button in the cell toolbar to run a cell too.
* When a cell is active you'll see a small menu in the upper-right with extra options.
* Press <kbd style="background-color: white; color: black; white-space: nowrap;">esc</kbd> to enter Command Mode and <kbd style="background-color: white; color: black; white-space: nowrap;">Alt</kbd> + <kbd style="background-color: white; color: black; white-space: nowrap;">↑ Up</kbd> or <kbd style="background-color: white; color: black; white-space: nowrap;">↓ Down</kbd> to move, <kbd style="background-color: white; color: black; white-space: nowrap;">c</kbd> to copy, or <kbd style="background-color: white; color: black; white-space: nowrap;">d</kbd> to delete cells. <nobr><q>**NOTE:** *The cell's blue outline disappears when you are in Command Mode.*</q></nobr>
* To navigate between cells, use the <kbd style="background-color: white; color: black; white-space: nowrap;">↑ Up</kbd> and <kbd style="background-color: white; color: black; white-space: nowrap;">↓ Down</kbd> arrow keys.
* You can also use the <kbd style="background-color: white; color: black; white-space: nowrap;">A</kbd> and <kbd style="background-color: white; color: black; white-space: nowrap;">B</kbd> keys to add new cells above or below the current cell.

**What to do if a cell doesn't run**

__1)__ Check the Python interpreter at the top of Visual Studio Code or your Codespace
<br>
__2)__ Click the kernel selector (if it says *Select Kernel*)
<br>
__3)__ Choose the <code style="background-color: white; color: black">.venv</code> or a system interpreter (e.g., like <code style="background-color: white; color: black">.venv (Python 3.13.5)</code>, <code style="background-color: white; color: black">Python 3.13.5</code>, or a similar option.).

<img src="https://images.jointheleague.org/vscode/select_python.png" width="50%" height="auto">

After selecting the interpreter, re-run the cell, or ask your instructor if you are still confused and need help connecting the kernel.


</div>

### **Getting Started**

Although we have already worked with variables before in the past [lesson](../10_Turtles/50_Variables_and_Functions/10_Variables.ipynb) (click the link for a refresher), we will now dive deeper into how they work and examine the various data types available in Python.

When you're ready to begin, run the code block below.

In [ ]:
# Run Me!

# Create some variables with different types and assign them values
age_in_years = 12
bank_account = "None"
name = "Aishini"
favorite_colors = ["red", "purple", "pink"]
gender = "Girl"
school = "Pacific Trails Middle Schoool"

# Now print the values using \n to print the next value on a new line

# this was given: print(f"Name: {name}\nAge: {age}\nBank Account: {bank_account}\nColors: {favorite_colors}")
# I made the python below:
print(f"Name: {name}\nAge In Years: {age_in_years}\nGender: {gender}\nSchool: {school}\nFavorite Colors: {favorite_colors}")
print(f"Bank Account: {bank_account}")


Name: Aishini
Age In Years: 12
Gender: Girl
School: Pacific Trails Middle Schoool
Favorite Colors: ['red', 'purple', 'pink']
Bank Account: None


In [23]:
# Run Me!
age = 12
bank_account = "none"
name = 'Aishini'
colors = ["red", "taupe"]

print("Age    : ", type(age))
print("Bank   : ", type(bank_account))
print("Name   : ", type(name))
print("Colors : ", type(colors))

Age    :  <class 'int'>
Bank   :  <class 'str'>
Name   :  <class 'str'>
Colors :  <class 'list'>


Some common data types you'll see:

* <strong>Integers:</strong> A whole number without a decimal part, like `3`
* <strong>Floats:</strong> Real numbers that can have a decimal part, like `3.5`
* <strong>Strings:</strong> Text data that is <span title="An immutable object CANNOT be modified after it is created." style="cursor: help">**immutable**<svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span>, like `"Alice"`
* <strong>Lists:</strong> are ordered sequences that are <span title="A mutable object CAN be modified after it is created." style="cursor: help">**mutable**<svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span>, like `['red', 'blue']`

> **Note:** There are many other types in Python, like <strong>bool</strong> (for `True`/`False` values) or <strong>dict</strong> (for key-value pairs), but the four above are some of the most common. 

### **Addition vs Concatenation**

Different types support different operations — for example, `+` performs arithmetic on numbers but concatenation on strings.

Let's see what happens when we add two integers and when we add two strings:

In [27]:
# Run Me!

# Integers
x = 10
y = 20

# Add x and y together
print("x + y =", x + y, x * y)

# Strings
x = "10"
y = "20"

# Add x and y together
print("x + y =", x + y) # Watch what happens.

x + y = 30 200
x + y = 1020


When we add `10` and `20` as integers, we get `30`. But when we add them as strings, they are <span title="The operation of joining two or more strings end-to-end to form a single, new string." style="cursor: help;">**concatenated**<svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span> together as `"1020"`.

But what happens if you try to add an integer to a string?

In [ ]:
# Run Me!

# Assign x and y to different types
x = 10
y = "20"

# Now lets see what happens when we try to add them
print("x + y =", x + y)

Oh, that caused a <span style="cursor: help; font-family: monospace; color: #CD3131;" title="A common exception that occurs when an operation or function is applied to an object of an inappropriate or unexpected data type"><strong>TypeError</strong><svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span> because you can't add an integer to a string. You *can*, however, convert one to the other type first and then add them.

###  **Type Conversion**
You can use `int()` to turn a string into a number, or `str()` to turn a number into a string. This lets you combine them in the way you want.

In [31]:
# Converting between types

# Assign x and y to different types
x = 10
y = "20"

# Convert y to an integer, then add x and y together
print("Integer:", x + int(y))

# Convert x to a string, then add x and y together
print("String:", str(x) + y)

Integer: 30
String: 1020



> **Tip:** The conversion function has the same name as the type you want to convert to. For example, `int()` makes an integer, `str()` makes a string.

Here are some common math operators for numbers. Notice how `//` and `%` work together:

| Operator | Description | Example | Result |
|----------|-------------|---------|--------|
| $+$        | Addition    | $2$ $+$ $3$   | $5$      |
| $-$        | Subtraction | $5$ $-$ $2$   | $3$      |
| $*$        | Multiplication | $4$ $*$ $6$ | $24$     |
| $/$        | Division    | $11$ $/$ $4$  | $2.75$   |
| $//$       | Floor division | $11$ $//$ $4$ | $2$ |
| $\%$       | Modulo (*Remainder*) | $11$ $\%$ $4$ | $3$ |
| $**$       | Exponentiation (*Power*) | $2$ ** $3$ | $8$ |

### **Floor Division**

Floor division uses the `//` operator. It divides and then drops anything after the decimal point. This means it returns an integer if you use integers, or a float if you use floats.

For example, `11 // 4` gives `2` (because 4 goes into 11 two times), and `11.0 // 4.0` gives `2.0`.

In [ ]:
# Run Me!

# Regular division will result in a float even with integers
print(22 / 7)

# Floor division with floats will also result in a float
print(22.0 // 7.0)

# Floor division with integers will result in an integer
print(22 // 7)

3.142857142857143
3.0
3


What if you want to find the **remainder** after dividing? That's where the modulo operator (`%`) comes in!

#### **Modulo**

The <span title="An arithmetic operator that returns the remainder of the division of one number (the dividend) by another (the divisor)." style="cursor: help;">**Modulo**<svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span> operator (`%`) gives you the remainder after dividing one number by another. Think of it like a clock: when you count up and reach the divisor, you start back at zero.

A common use for modulo is checking if a number is even or odd:
- `4 % 2 == 0` means 4 is even (no remainder)
- `5 % 2 == 1` means 5 is odd (remainder is 1)

In general, if `n % divisor == 0`, then `n` is evenly divisible by that divisor.

In [45]:
# Run Me!

# Modulo Operator - Check if numbers are even or odd

print("4 % 2 ==", 4 % 2)
print("6 % 2 ==", 6 % 2)
print("5 % 2 ==", 5 % 2)
print("6 % 2 ==", 6 % 2)

4 % 2 == 0
6 % 2 == 0
5 % 2 == 1
6 % 2 == 0


### **Connecting Division and Modulo**

The floor division (`//`) and modulo (`%`) operators are closely related! Together, they let you break a number into two parts: the <span style="cursor: help;" title="The answer you get when you divide one number by another. For instance, in the equation: 10 ÷ 4 = 2.5, the quotient is 2.5"><strong>quotient</strong><svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span> and the <span style="cursor: help;" title="The whole number amount left over after performing integer division (how many whole times a number fits into another). For instance, in the equation 10 ÷ 4 = 2.5, the remainder is 2—because the divisor (4) goes into 10 twice (e.g., 4 * 2 = 8)—and 2 is left over."><strong>remainder</strong><svg style="width:18px;height:18px; vertical-align: middle; margin-left: 2px; margin-bottom: 3px;" viewBox="0 0 24 24"><path fill="currentColor" d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M11,16.5V11.5H13V16.5H11M11,9.5V7.5H13V9.5H11Z"/></svg></span>.

You can always get back the original number using this formula:
$dividend = (quotient * divisor) + remainder$

In [49]:
# Run Me!

# Define numbers
dividend = 11
divisor = 4

# Calculate Quotient and Remainder
quotient = dividend // divisor   # Whole number result
remainder = dividend % divisor   # Leftover amount

# Print results
print(f"If you divide {dividend} by {divisor}, the quotient is {quotient}, and the remainder is {remainder}.")

# Verification: (Quotient * Divisor) + Remainder = Dividend
print(f"\nVerification: ({quotient} * {divisor}) + {remainder} = {dividend}")
print(f"\nResult: {quotient * divisor + remainder}")

If you divide 11 by 4, the quotient is 2, and the remainder is 3.

Verification: (2 * 4) + 3 = 11

Result: 11


As shown above, when you divide `11` by `4`, the quotient is `2` and the remainder is `3`.

Let's see how this works for a range of numbers:

In [52]:
# Run Me!

# Example of the Modulo Operator

for i in range(1,11):
    print(f"Floor Division: {i} // 3 = {i // 3} | Modulo: {i} % 3 = {i % 3} | Rebuild = {i // 3} * 3 + {i % 3} = {i // 3 * 3 + i % 3}")

Floor Division: 1 // 3 = 0 | Modulo: 1 % 3 = 1 | Rebuild = 0 * 3 + 1 = 1
Floor Division: 2 // 3 = 0 | Modulo: 2 % 3 = 2 | Rebuild = 0 * 3 + 2 = 2
Floor Division: 3 // 3 = 1 | Modulo: 3 % 3 = 0 | Rebuild = 1 * 3 + 0 = 3
Floor Division: 4 // 3 = 1 | Modulo: 4 % 3 = 1 | Rebuild = 1 * 3 + 1 = 4
Floor Division: 5 // 3 = 1 | Modulo: 5 % 3 = 2 | Rebuild = 1 * 3 + 2 = 5
Floor Division: 6 // 3 = 2 | Modulo: 6 % 3 = 0 | Rebuild = 2 * 3 + 0 = 6
Floor Division: 7 // 3 = 2 | Modulo: 7 % 3 = 1 | Rebuild = 2 * 3 + 1 = 7
Floor Division: 8 // 3 = 2 | Modulo: 8 % 3 = 2 | Rebuild = 2 * 3 + 2 = 8
Floor Division: 9 // 3 = 3 | Modulo: 9 % 3 = 0 | Rebuild = 3 * 3 + 0 = 9
Floor Division: 10 // 3 = 3 | Modulo: 10 % 3 = 1 | Rebuild = 3 * 3 + 1 = 10


### **Challenge**

Imagine you have some candy bars and you have to share them evenly amongst a group of kids. Adjust the variables and calculations below to determine how many candy bars each kid gets and how many are left over.

**Objectives:** 
- Set `kids` to the number of kids
- Set `candy_bars` to the number of candy bars you have

**Hints:** 
- Use `//` to find how many each kid gets
- Use `%` to find the remainder of candy bars

In [26]:
# Test yourself

# this was mainly given, I just filled out numbers and a few things: 

# kids = 5       # Set the number of kids
# candy_bars = 10 # Set the number of candy bars

# candy_per_kid = candy_bars//kids   # Calculate the number of candy bars each kid gets
# print("Each kid gets", candy_per_kid, "candy bars")

# candy_left_over = candy_per_kid % candy_bars # Calculate the number of candy bars left over
# print("There are", candy_left_over, "candy bars left over")

# Everything below I coded myself

# assigning variables:
people = 17
tables = 7
chairs = 15

# creating equations:
chairs_per_table = chairs//tables
people_without_seats = people % chairs
people_with_seats = chairs

# printing the final calculations, and story/background
print("People, Chairs, and Tables:")
print("At a small family restaurant, there is", people, "customers.")
print("Due to financial difficulties, they could only afford a few amount of chairs.")
print("They also had the policy, that only one person could sit down per chair.")
print("The restaurant had only had", chairs, "chairs available for usage.")
print("There must be the same number of chairs per table.\nSo, there will be", chairs_per_table, "chairs per table.")
print("Unfortunately, not everyone will have a seat.\nThere will be", people_without_seats, "people without a seat.")
print("While there is 2 people without a seat, still", chairs, "people have a place to sit.")
print("So overall, there will be", chairs_per_table, "chairs per table.")
print("With", people_with_seats, "people with a seat, and", people_without_seats, "people without seats.")



People, Chairs, and Tables:
At a small family restaurant, there is 17 customers.
Due to financial difficulties, they could only afford a few amount of chairs.
They also had the policy, that only one person could sit down per chair.
The restaurant had only had 15 chairs available for usage.
There must be the same number of chairs per table.
So, there will be 2 chairs per table.
Unfortunately, not everyone will have a seat.
There will be 2 people without a seat.
While there is 2 people without a seat, still 15 people have a place to sit.
So overall, there will be 2 chairs per table.
With 15 people with a seat, and 2 people without seats.


<div class="alert alert-block alert-info">
<strong>Reminder:</strong> Check in your code. If you need a refresher, review the <a href="https://curriculum.jointheleague.org/howto/checkin_restart.html"><strong>Check in Code and Restart Codespaces</strong></a> guide or circle back to the previous <a href="../10_Turtles/20_Introducing_Tina/40_Check_In_Your_Code.ipynb"><strong>Check In Your Code</strong></a> lesson.
</div>